pkg/cosign/env/env.go

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package env
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+)
+
+type Variable string
+
+type VariableOpts struct {
+	Description string
+	Expects     string
+	Sensitive   bool
+}
+
+func (v Variable) String() string {
+	return string(v)
+}
+
+const (
+	VariableExperimental     Variable = "COSIGN_EXPERIMENTAL"
+	VariableDockerMediaTypes Variable = "COSIGN_DOCKER_MEDIA_TYPES"
+	VariablePassword         Variable = "COSIGN_PASSWORD"
+	VariablePKCS11Pin        Variable = "COSIGN_PKCS11_PIN"
+	VariablePKCS11ModulePath Variable = "COSIGN_PKCS11_MODULE_PATH"
+	VariableRepository       Variable = "COSIGN_REPOSITORY"
+)
+
+var (
+	environmentVariables = map[Variable]VariableOpts{
+		VariableExperimental: {
+			Description: "enables experimental cosign features",
+			Expects:     "1 if experimental features should be enabled (0 by default)",
+			Sensitive:   false,
+		},
+		VariableDockerMediaTypes: {
+			Description: "to be used with registries that do not support OCI media types",
+			Expects:     "1 to fallback to legacy OCI media types equivalents (0 by default)",
+			Sensitive:   false,
+		},
+		VariablePassword: {
+			Description: "overrides password inputs with this value",
+			Expects:     "string with a password (asks on stdin by default)",
+			Sensitive:   true,
+		},
+		VariablePKCS11Pin: {
+			Description: "to be used if PKCS11 PIN is not provided",
+			Expects:     "string with a PIN",
+			Sensitive:   true,
+		},
+		VariablePKCS11ModulePath: {
+			Description: "is PKCS11 module-path",
+			Expects:     "string with a module-path",
+			Sensitive:   false,
+		},
+		VariableRepository: {
+			Description: "can be used to store signatures in an alternate location",
+			Expects:     "string with a repository",
+			Sensitive:   false,
+		},
+	}
+)
+
+func mustRegisterEnv(name Variable) {
+	if _, ok := environmentVariables[name]; !ok {
+		panic(fmt.Sprintf("environment variable %q is not registered in pkg/cosign/env", name.String()))
+	}
+	if !strings.HasPrefix(name.String(), "COSIGN_") {
```

Comment:
Welp, now that I've asked you to Enforce the `COSIGN_` prefix we might want to start doing `SIGSTORE_` and `TUF_` variables. Can you allowlist all three prefixes? You could also get rid of this altogether in favor of the "is in `environmentVariables`" check

I think in the medium-to-long-term we probably want to do something similar to this pattern in the upstream sigstore/TUF repositories, and then we could really go back to only having `COSIGN_` variables (we'd call `sigstore/pkg/env.LookupEnv` where needed).

Reply:
I went with some in-between approach here. I added a new field to `VariableOpts` -- `External`. If variable is not external (`External: false`), then it must start with `COSIGN_`, otherwise we don't care.

Reply:
nice!